### DISCUSSION TITLE
Best way to parse ascii letters and numbers

### DISCUSSION THREAD

**Discussion by DamianReeves** [j-mie6/Parsley#224] [Q&A] [answered]
I am implementing a parser for [the package url (purl) spec](https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst) and in looking at the type section, I need to deal with ASCII letters and numbers. What is the best way to construct the parser for `type`?

> type:
>
>    The package type is composed only of ASCII letters and numbers, '.', '+' and '-' (period, plus, and dash)
>    The type cannot start with a number
>    The type cannot contains spaces
>    The type must NOT be percent-encoded
>    The type is case insensitive. The canonical form is lowercase

**Comment by j-mie6** [comment, marked as answer]
It's not clear what _exactly_ it means by "not be percent encoded"; whether that's supposed to be "no percent-encoded characters in the name" or what. But basically, it seems like a direct translation is probably best:

```scala
val start = oneOf('a' to 'z') | oneOf('A' to 'Z') | oneOf('.', '+', '-')
val letters = oneOf('0' to '9') | start
```

(you could use `digit`/`letter`, but that would be more than _just_ ASCII in this case; depends on how strict you want to be. Or you could build a set of all the letters and use a single `oneOf`, which would be more efficient in this case). Then you want to stitch them together, which can probably be done with `span` for simplicity:

```scala
val typePurl = (start ~> many(letters)).span.map(_.toLowerCase)
```
That will give you a canonicalisation to lowercase strings (I might have gotten the specific method wrong from `Char`, but you get the idea.

Then the percent-encoded bit should be trivially satisfied by not allowing for any `%` in the name... if I understand it correctly? Otherwise, you could finish it off with a `filter` on the `typePurl` to check for whatever is needed inside?

**Reply by DamianReeves** [reply]
Thanks so much for the quick response. This is awesome and makes sense.
